### PR TITLE
Extract message crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_program::message` needs to be pulled out

#### Summary of Changes
- Move to its own crate and re-export with deprecation
- Move AddressLookupTableAccount to this crate
- Move CompiledInstruction to this crate
- Copy ALL_IDS from the sysvar module to avoid a dependency. This code is deprecated anyway so the stuff we are copying should not change.
- Inline the NONCED_TX_MARKER_IX_INDEX from solana_nonce to avoid a dependency
- Make serde and bincode optional in this crate
- Remove thiserror from the new crate
- Add doc_auto_cfg

This uses the [solana-system-interface](https://github.com/solana-program/system) crate so that needs to be published before this can be merged (update: solana-system-interface has been published)

### CODE REVIEW

File: sdk/message/src/legacy.rs [hunk 1]
```diff
@@ -5,30 +5,58 @@
...
+        sysvar::fees::id(),
+        sysvar::recent_blockhashes::id(),
+        sysvar::rent::id(),
+        sysvar::rewards::id(),
+        sysvar::slot_hashes::id(),
+        sysvar::slot_history::id(),
+        sysvar::stake_history::id(),
+        sysvar::instructions::id(),
+    ];
+}
```
Comment (joncinque) [inline sdk/message/src/legacy.rs:52]: This makes sense to me. To be safe, can you add a test to make sure that `ALL_IDS == solana_sysvar::ALL_IDS`?

File: sdk/message/src/sanitized.rs [hunk 2]
```diff
@@ -323,20 +326,23 @@ impl SanitizedMessage {
             })
     }
 
+    #[cfg(feature = "bincode")]
```
Comment (joncinque) [inline sdk/message/src/sanitized.rs:337]: These feature annotations pretty much give the full roadmap about how we can remove bincode from everything -- it's going to be really useful in the future :pray:

File: sdk/message/src/versions/v0/loaded.rs [hunk 3]
```diff
@@ -270,8 +272,6 @@ mod tests {
 
     #[test]
     fn test_is_writable() {
-        solana_logger::setup();
```
Comment (joncinque) [inline sdk/message/src/versions/v0/loaded.rs:273]: :partying_face:

File: sdk/transaction-error/src/lib.rs
Comment (joncinque) [inline sdk/transaction-error/src/lib.rs:1]: Were these changes intended?

**Review by joncinque — COMMENTED** [review]
Looks really good, just one suggestion to add a test and to revert some changes that might have snuck in. Is this the last of the painful crate extractions from solana-program?

File: sdk/transaction-error/src/lib.rs
Comment (kevinheavey) [reply sdk/transaction-error/src/lib.rs:1]: yes, for some reason  the CI only complained about an unused import in this PR. Maybe the warning wasn't causing any CI to fail. But the warning is real

Comment (joncinque) [reply sdk/transaction-error/src/lib.rs:1]: Gotcha, makes sense -- it's unused in `target_os = "solana"`, so if you run `./cargo-build-sbf --manifest-path sdk/transaction-error/Cargo.toml`, you'll see that ~~error~~ warning. I'll put in a separate PR for that if that's ok

Comment (joncinque) [reply sdk/transaction-error/src/lib.rs:1]: #3742 fixes that and another thing

**Review by joncinque — DISMISSED** [review]
Looks great, thanks! @yihau can you accept `solana-message`?

**Review by joncinque — DISMISSED** [review]

**Comment by joncinque** [comment]
Sorry, can you rebase once more to pick up https://github.com/anza-xyz/agave/pull/3774 ?

**Comment by kevinheavey** [comment]
Need to figure out some doctest dependency issues

**Comment by joncinque** [comment]
> Need to figure out some doctest dependency issues

There's another issue with the cargo build-sbf tests -- since they're using a local solana-program without patching the deps for solana-system-interface, the build pulls in two versions of solana-pubkey / solana-instruction / solana-decode-error. I can push up a fix for this if you like

**Comment by kevinheavey** [comment]
Oh yes please, maybe that's it

**Comment by joncinque** [comment]
I think we're looking good! Tagged @febo for an approval since I can't merge it :upside_down_face:

**Review by febo — DISMISSED** [review]
Looks great! There are some CI tasks failing.

**Comment by joncinque** [comment]
Looks like there's a missing dependency on wasm-bindgen. But even after fixing that, there's multiply defined symbols for the system instructions between `solana-system-interface` and `sdk/program/src/system_instruction.rs` -- we'll need to remove those first and change solana-program to re-export all of those.

**Comment by febo** [comment]
> Looks like there's a missing dependency on wasm-bindgen. But even after fixing that, there's multiply defined symbols for the system instructions between `solana-system-interface` and `sdk/program/src/system_instruction.rs` -- we'll need to remove those first and change solana-program to re-export all of those.

Would you like me do have a look at it?

**Comment by joncinque** [comment]
> Would you like me do have a look at it?

Yes please that'd be great, whenever you have a moment

**Review by joncinque — APPROVED** [review]
Looks great, thanks for the quick rebase!
